### PR TITLE
[SPARK-13815][MLlib] Provide better Exception messages in Pipeline load methods

### PR DESCRIPTION
**Pull request by joan38**
## What changes were proposed in this pull request?

Provides better Exception messages in Pipeline load methods when load empty metadata.
## How was this patch tested?

Manual tests

### CODE REVIEW

File: mllib/src/main/scala/org/apache/spark/ml/util/ReadWrite.scala [hunk 1]
```diff
@@ -282,7 +282,10 @@ private[ml] object DefaultParamsReader {
    */
   def loadMetadata(path: String, sc: SparkContext, expectedClassName: String = ""): Metadata = {
     val metadataPath = new Path(path, "metadata").toString
-    val metadataStr = sc.textFile(metadataPath, 1).first()
+    val metadataStr = try sc.textFile(metadataPath, 1).first() catch {
+      case _: UnsupportedOperationException =>
+        throw new IllegalArgumentException(s"Empty metadata for path $path")
+    }
```
Comment (joan38) [inline mllib/src/main/scala/org/apache/spark/ml/util/ReadWrite.scala]: Or
```scala
    val textFile = sc.textFile(metadataPath, 1)
    if (textFile.isEmpty()) throw new IllegalArgumentException(s"Empty metadata for path $path")
    val metadataStr = textFile.first()
```
But I thought calling twice `take(1)` threw `isEmpty()` and `first()` would be a waist.

File: mllib/src/main/scala/org/apache/spark/ml/util/ReadWrite.scala [hunk 2]
```diff
@@ -282,7 +282,10 @@ private[ml] object DefaultParamsReader {
    */
   def loadMetadata(path: String, sc: SparkContext, expectedClassName: String = ""): Metadata = {
     val metadataPath = new Path(path, "metadata").toString
-    val metadataStr = sc.textFile(metadataPath, 1).first()
+    val metadataStr = try sc.textFile(metadataPath, 1).first() catch {
+      case _: UnsupportedOperationException =>
+        throw new IllegalArgumentException(s"Empty metadata for path $path")
```
Comment (joan38) [inline mllib/src/main/scala/org/apache/spark/ml/util/ReadWrite.scala]: Let me know if you have any better error message.

**Comment by srowen** [comment]
I don't think this is a great fix, since it's just catching a general exception (which could be triggered by more than just the case in question) and it isn't changing the exception for other cases. I think this is probably not worth changing, as the error for this case already suggests that something is empty, but is being treated as if it's not.

**Comment by joan38** [comment]
> I don't think this is a great fix, since it's just catching a general exception (which could be triggered by more than just the case in question)

I don't agree with this if we use the 2nd suggested version:
```scala
    val textFile = sc.textFile(metadataPath, 1)
    if (textFile.isEmpty()) throw new IllegalArgumentException(s"Empty metadata for path $path")
    val metadataStr = textFile.first()
```

> it isn't changing the exception for other cases.

Agreed

> I think this is probably not worth changing, as the error for this case already suggests that something is empty, but is being treated as if it's not.

Fair enough

**Pull request closed without merging**